ante/fee.go

```diff
@@ -186,85 +175,59 @@ func (dfd DeductFeeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bo
 	if !feeTx.GetFee().IsZero() {
+		nativeFees := sdk.Coins{}
+		xatpFees := sdk.Coins{}
 
-		if isXpla {
-			err = DeductFees(dfd.bankKeeper, ctx, deductFeesFromAcc, feeTx.GetFee())
-			if err != nil {
-				return ctx, err
-			}
-
-		} else {
+		xatpPayer := dfd.xatpKeeper.GetPayer(ctx)
 
-			ctx = ctx.WithGasMeter(sdk.NewGasMeter(sdk.Gas(dfd.smartQueryGasLimit)))
-			for _, coin := range fee {
-
-				denom := coin.Denom
-
-				ratioDec, err := dfd.xatpKeeper.GetFeeInfoFromXATP(ctx, denom)
-				if err != nil {
-					return ctx, err
-				}
-
-				var xplaDecimal sdk.Dec
-				if denom != xplatypes.DefaultDenom {
-
-					minGasPrices, err := sdk.ParseDecCoins(dfd.MinGasPrices)
-					if err != nil {
-						return ctx, err
-					}
-
-					var decimal int64
-					for _, gp := range minGasPrices {
-
-						if gp.Denom == xplatypes.DefaultDenom {
-							decimal = int64(len(gp.Amount.RoundInt().String()))
-						}
-					}
-
-					xplaDecimal = sdk.NewDecFromBigInt(new(big.Int).Mul(big.NewInt(1), new(big.Int).Exp(big.NewInt(10), big.NewInt(decimal), nil)))
+		for _, fee := range feeTx.GetFee() {
+			if xatpPayer == "" {
+				break
+			}
 
-					err = dfd.xatpKeeper.PayXATP(ctx, deductFeesFrom, denom, coin.Amount.String())
-					if err != nil {
-						return ctx, err
-					}
-				}
+			xatp, found := dfd.xatpKeeper.GetXatp(ctx, fee.Denom)
+			if !found {
+				nativeFees = nativeFees.Add(fee)
+				continue
+			}
 
-				xatpPayer := dfd.xatpKeeper.GetPayer(ctx)
-				XatpPayerAcc, err := sdk.AccAddressFromBech32(xatpPayer)
-				if err != nil {
-					return ctx, err
-				}
+			err := dfd.xatpKeeper.PayXATP(ctx, deductFeesFrom, xatp.Denom, fee.Amount.String())
+			if err != nil {
+				return ctx, err
+			}
 
-				accExists := dfd.ak.(authkeeper.AccountKeeper).HasAccount(ctx, XatpPayerAcc)
-				if !accExists {
-					return ctx, sdkerrors.Wrapf(sdkerrors.ErrUnknownAddress, "XATP payer account: %s not found", XatpPayerAcc)
-				}
+			ratioDec, err := dfd.xatpKeeper.GetFeeInfoFromXATP(ctx, xatp.Denom)
+			if err != nil {
+				return ctx, err
+			}
 
-				deductFeesFromAcc0 := dfd.ak.GetAccount(ctx, XatpPayerAcc)
+			feeAmount := sdk.NewDecFromIntWithPrec(fee.Amount, int64(xatp.Decimals))
+			defaultFeeAmountDec := feeAmount.Quo(ratioDec)
+			defaultFeeAmount := defaultFeeAmountDec.Mul(sdk.NewDecWithPrec(10, xplatypes.DefaultDenomPrecision)).Ceil().BigInt()
+			xatpFees = xatpFees.Add(sdk.NewCoin(xplatypes.DefaultDenom, sdk.NewIntFromBigInt(defaultFeeAmount)))
+		}
 
-				xplaValue := sdk.NewDecFromInt(coin.Amount).Quo(ratioDec)
-				xplaValue = xplaValue.Mul(xplaDecimal)
-				xplaFee := sdk.NewCoins(sdk.NewCoin(xplatypes.DefaultDenom, xplaValue.Ceil().RoundInt()))
+		if !nativeFees.Empty() {
+			err = DeductFees(dfd.bankKeeper, ctx, deductFeesFromAcc, nativeFees)
+			if err != nil {
+				return ctx, err
+			}
+		}
 
-				err = DeductFees(dfd.bankKeeper, ctx, deductFeesFromAcc0, xplaFee)
-				if err != nil {
-					return ctx, err
-				}
+		if !xatpFees.Empty() {
```

Comment:
XATP only exeutes on the node which received the Tx

I added a debug print line as https://github.com/xpladev/xpla/commit/0e53329a7c705b17f186cf10cd3eff83c9529b8e
(one common line vs only XATP Tx executes)

In the integration tests, all tests with the common fee prints the common print
But in the case of XATP tx, the tx is only executed in the node which is received the tx


Printed the event and it executes well

```
events:
_contract_address : xpla14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s525s0h
_contract_address : xpla14hj2tavq8fpesdwxxcu44rty3hh90vhujrvcmstl4zr3txmfvw9s525s0h
action : transfer
from : xpla1z2k85n48ydfvzslrugwzl4j2u7vtdyf3xvucmc
to : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 360000
spender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 1000000000000000000axpla
receiver : xpla17xpfvakm2amg962yls6f84z3kell8c5lw7muqw
amount : 1000000000000000000axpla
recipient : xpla17xpfvakm2amg962yls6f84z3kell8c5lw7muqw
sender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
amount : 1000000000000000000axpla
sender : xpla15wdvvke4munrlchecfhdphhkfdw4pzmrpdejnd
```

Reply:
Classified as a bad transaction in ante/fee.go::Mempoolhandler, L103 due to low fee

Reply:
FYI

Also fails that the simulation does even not raise simulation error with XPLA fee
https://github.com/xpladev/xpla/blob/test/xatp/integration_test/integration_test.go#L788

Reply:
For simulation, Tx must be in sync mode. If you use sync mode you will get an error in TxResponse. It is not a broadcastTx error.

https://github.com/xpladev/xpla/blob/test/xatp/integration_test/wallet_test.go#L242-L250

Reply:
Solved as below in the testing side and works well 👍 

```go
if currtestingenv := os.Getenv("GOLANG_TESTING"); currtestingenv != "true" {
res, err := client.BroadcastTx(context.Background(), &txtype.BroadcastTxRequest{
	TxBytes: txBytes,
	Mode:    mode,
})

if err != nil {
	err = errors.Wrap(err, "broadcastTx")
	return "", err
}

// if the tx works well in the sync mode, it returns as "[]"
if res.TxResponse.RawLog != "[]" {
	return "", fmt.Errorf(res.TxResponse.RawLog)
}
```